pandas/tests/series/methods/test_drop_duplicates.py

```diff
@@ -67,76 +67,164 @@ def test_drop_duplicates_no_duplicates(any_numpy_dtype, keep, values):
         # Test case 1
+        cat_array = np.array([1, 2, 3, 4, 5], dtype=np.dtype(dtype))
+
         input1 = np.array([1, 2, 3, 3], dtype=np.dtype(dtype))
+
         tc1 = Series(Categorical(input1, categories=cat_array, ordered=ordered))
-        if dtype == "datetime64[D]":
-            # pre-empty flaky xfail, tc1 values are seemingly-random
-            if not (np.array(tc1) == input1).all():
-                pytest.xfail(reason="GH#7996")
+        return tc1
+
+    def _maybe_xfail_tc(self, tc, request):
+        warn = None
+        if tc.cat.categories.dtype.kind == "M":
+            if len(tc) == 4:
+                # This is cat_series1
+                input_arr = np.array([1, 2, 3, 3], dtype=np.dtype("datetime64[D]"))
+            else:
+                # This is tc2
+                input_arr = np.array(
+                    [1, 2, 3, 5, 3, 2, 4], dtype=np.dtype("datetime64[D]")
+                )
+
+            if not (np.array(tc) == input_arr).all():
+                mark = pytest.mark.xfail(
+                    reason="GH#7996 tc1/tc2 values are seemingly-random",
+                    raises=AssertionError,
+                )
+                request.node.add_marker(mark)
+
+                # Also warn about Timestamp vs date deprecation
+                warn = FutureWarning
+        return warn
+
+    def test_drop_duplicates_categorical_non_bool(self, cat_series1, request):
+        tc1 = cat_series1
+        warn = self._maybe_xfail_tc(tc1, request)
 
         expected = Series([False, False, False, True])
-        tm.assert_series_equal(tc1.duplicated(), expected)
-        tm.assert_series_equal(tc1.drop_duplicates(), tc1[~expected])
-        sc = tc1.copy()
-        return_value = sc.drop_duplicates(inplace=True)
-        assert return_value is None
-        tm.assert_series_equal(sc, tc1[~expected])
+
+        with tm.assert_produces_warning(warn):
+            result = tc1.duplicated()
+            tm.assert_series_equal(result, expected)
+
+            result = tc1.drop_duplicates()
+            tm.assert_series_equal(result, tc1[~expected])
+
+            sc = tc1.copy()
+            return_value = sc.drop_duplicates(inplace=True)
+            assert return_value is None
+            tm.assert_series_equal(sc, tc1[~expected])
+
+    def test_drop_duplicates_categorical_non_bool_keeplast(self, cat_series1, request):
+        tc1 = cat_series1
+        warn = self._maybe_xfail_tc(tc1, request)
 
         expected = Series([False, False, True, False])
-        tm.assert_series_equal(tc1.duplicated(keep="last"), expected)
-        tm.assert_series_equal(tc1.drop_duplicates(keep="last"), tc1[~expected])
-        sc = tc1.copy()
-        return_value = sc.drop_duplicates(keep="last", inplace=True)
-        assert return_value is None
-        tm.assert_series_equal(sc, tc1[~expected])
+
+        with tm.assert_produces_warning(warn):
+            result = tc1.duplicated(keep="last")
```

Comment:
i find these warnings user facing is odd

Reply:
Looks like these can _almost_ be removed, but for a weird interaction with the flaky xfail in this test, which I _think_ i can track down to the Categorical constructor, will take a swing at separately.

Reply:
Fixed by #38791